packages/flutter_driver/lib/src/common/wait.dart

```diff
@@ -0,0 +1,185 @@
+
+  /// Factory constructor to parse a [NoPendingFrameCondition] instance from the
+  /// given JSON map.
+  factory NoPendingFrameCondition.fromJson(Map<String, dynamic> json) {
+    if (json['conditionName'] != 'NoPendingFrameCondition')
+      throw SerializationException(
+          'Not a NoPendingFrameCondition json string.');
+    return const NoPendingFrameCondition();
+  }
+
+  @override
+  bool get condition => !SchedulerBinding.instance.hasScheduledFrame;
+
+  @override
+  Future<void> wait() async {
+    if (condition) {
+      return;
+    } else {
+      do {
+        await SchedulerBinding.instance.endOfFrame;
+      } while (!condition);
+    }
+    assert(condition);
+  }
+
+  @override
+  Map<String, String> serialize() => <String, String>{
+        'conditionName': 'NoPendingFrameCondition',
+      };
+}
+
+/// A combined condition that waits until all the given [conditions] are met.
+class CombinedCondition implements WaitCondition {
+  /// Creates a [CombinedCondition] instance with the given list of
+  /// [conditions].
+  const CombinedCondition(this.conditions);
+
+  /// Factory constructor to parse a [CombinedCondition] instance from the given
+  /// JSON map.
+  factory CombinedCondition.fromJson(Map<String, dynamic> jsonMap) {
+    if (jsonMap['conditionName'] != 'CombinedCondition')
+      throw SerializationException('Not a CombinedCondition json string.');
+    CombinedCondition combinedCondition;
+    if (jsonMap['conditions'] == null) {
+      combinedCondition = const CombinedCondition(<WaitCondition>[]);
+    } else {
+      final List<WaitCondition> conditions = <WaitCondition>[];
+      for (Map<String, dynamic> condition
+          in json.decode(jsonMap['conditions'])) {
+        conditions.add(WaitConditionDecoder.fromJson(condition));
+      }
+      combinedCondition = CombinedCondition(conditions);
+    }
+    return combinedCondition;
+  }
+
+  /// A list of conditions it waits for.
+  final List<WaitCondition> conditions;
+
+  @override
+  bool get condition =>
+      conditions.every((WaitCondition condition) => condition.condition);
+
+  @override
+  Future<void> wait() async {
+    do {
+      for (WaitCondition condition in conditions) {
+        if (condition != null) {
+          await condition.wait();
+        }
+      }
+    } while (!condition);
+    assert(condition);
+  }
+
+  @override
+  Map<String, String> serialize() {
+    final Map<String, String> jsonMap = <String, String>{
+      'conditionName': 'CombinedCondition'
+    };
+    final List<Map<String, String>> jsonConditions = <Map<String, String>>[];
+    for (WaitCondition condition in conditions) {
+      if (condition != null) {
+        jsonConditions.add(condition.serialize());
+      }
+    }
+    jsonMap.putIfAbsent('conditions', () => json.encode(jsonConditions));
+    return jsonMap;
+  }
+}
+
+/// A JSON decoder that parses JSON map to a [WaitCondition] or its subclass.
+class WaitConditionDecoder {
```

Comment:
Can this be private?

Reply:
Not really. It's also used in the `WaitForCondition.deserialize` in the find.dart file.

Reply:
Makes it private now.